src/models/geograph.py

```diff
@@ -66,50 +305,135 @@ def _dataframe_to_graph(
         Args:
             df (gpd.GeoDataFrame): GeoDataFrame containing polygon objects from
             a shape file.
-            attributes (Optional[List[str]], optional): columns of gdf that are
-            added to nodes of graph as attributes. Defaults to None.
+            attributes (Optional[List[str]], optional): columns of df that are
+            added to nodes of graph as attributes. Defaults to None,
+            which will cause all attributes to be added.
+            tolerance (float, optional): Adds edges between neighbours that are
+            at most `tolerance` metres apart. Defaults to 0.
+
+        Raises:
+            ValueError: If `tolerance` < 0.
+
+        Returns:
+            STRtree: The Rtree object used to build the graph.
         """
+        if tolerance < 0.0:
+            raise ValueError("`tolerance` must be greater than 0.")
         # If no attribute list is given, all
         # columns in df are used
         if attributes is None:
             attributes = df.columns.tolist()
+        elif not set(attributes).issubset(set(df.columns)):
```

Comment:
Super neat! Haven't seen that before but I like it very much! 🥇 😍
Is the second `set(...)` needed?

Reply:
Same thought when I saw this: very neat!

Reply:
I just tested and the second actually isn't needed, so I'll remove it.